<table>
        <td>
        <td><img src="https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/logo_ECI.png" width="400"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue"> DataFrames </p> Ejercicio </p> Funciones en Python  </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Diplomado Ciencia de Datos </p></tp>
            <tp><p style="font-size:115%;text-align:center">2026-1</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>


# <FONT SIZE=5 COLOR="blue"> Contenidos del Notebook </FONT>

En este notebook encontrará:

- Repaso de elementos de exploración de Dataframes usando Pandas

- Funciones en Python

```python
def function(x):
  ....
  return ...
```






# <FONT SIZE=5 COLOR="purple"> 1. Ejercicio de Exploración y Limpieza de los Datos </FONT>

**Contexto de los datos**

Este conjunto de datos contiene información financiera y demográfica de individuos relacionada con el uso y posesión de tarjetas de crédito. El objetivo principal es analizar los factores asociados a la tenencia de una tarjeta de crédito y estudiar el comportamiento financiero de los clientes.

Las variables incluyen características personales (edad, ingresos, número de dependientes), características laborales (trabajador independiente o no), situación patrimonial (propietario de vivienda) e información crediticia (historial de reportes negativos, número de tarjetas y cuentas activas).

Este conjunto de datos es ampliamente utilizado en cursos de estadística, minería de datos y aprendizaje automático para ilustrar técnicas de análisis exploratorio, filtrado de datos, agrupamiento (`groupby`), visualización y modelos de clasificación.

La variable de mayor interés suele ser `card`, que indica si una persona posee o no una tarjeta de crédito. A partir de las demás variables, es posible identificar patrones que ayuden a explicar o predecir dicha condición.


| Variable | Traducción | Descripción |
|-----------|------------|-------------|
| `card` | Tarjeta | Indica si la persona posee una tarjeta de crédito (sí/no). |
| `reports` | Reportes negativos | Número de reportes negativos o incidentes en el historial crediticio. |
| `age` | Edad | Edad de la persona (años). |
| `income` | Ingreso | Ingreso anual del individuo o hogar. |
| `share` | Proporción de gasto | Proporción del ingreso destinada a gastos con tarjeta. |
| `expenditure` | Gasto | Gasto mensual realizado con tarjeta de crédito. |
| `owner` | Propietario de vivienda | Indica si posee vivienda propia (sí/no). |
| `selfemp` | Trabajador independiente | Indica si es trabajador independiente o autónomo (sí/no). |
| `dependents` | Dependientes | Número de personas dependientes económicamente. |
| `months` | Meses de historial | Meses de historial crediticio o de permanencia en el sistema financiero. |
| `majorcards` | Tarjetas principales | Número de tarjetas de crédito principales que posee. |
| `active` | Cuentas activas | Número de cuentas de crédito activas. |

link de los datos



```python
https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/credit_diplomado.csv
```



Iniciaremos importando la librería *pandas*

In [2]:
# importamos las librerías
import pandas as pd
import numpy as np

In [ ]:
#na_values=" "

In [3]:
# cargamos los datos crédito
credito = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/credit_diplomado.csv")
credito.head()

,Trajeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
0,yes,0,37.66667,4.5200,0.033270,124.983300,yes,no,3,54,1,12.0
1,yes,0,33.25,2.4200,0.005217,9.854167,no,no,3,34,1,13.0
2,yes,0,33.66667,4.5000,0.004156,15.000000,yes,no,4,5o,1,5.0
3,yes,0,30.5,2.5400,0.065214,137.869200,no,no,0,25,1,7.0
4,yes,0,32.16667,9.7867,0.067051,546.503300,yes,no,2,64,1,5.0


A continuación haremos una exploración rápida de los datos para identificar posibles inconsistencia y poderlas solucionar.

In [4]:
# los primeros datos
credito.head()

,Trajeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
0,yes,0,37.66667,4.5200,0.033270,124.983300,yes,no,3,54,1,12.0
1,yes,0,33.25,2.4200,0.005217,9.854167,no,no,3,34,1,13.0
2,yes,0,33.66667,4.5000,0.004156,15.000000,yes,no,4,5o,1,5.0
3,yes,0,30.5,2.5400,0.065214,137.869200,no,no,0,25,1,7.0
4,yes,0,32.16667,9.7867,0.067051,546.503300,yes,no,2,64,1,5.0


observamos algunas variables numéricas y otras categóricas

In [ ]:
# últimos registros a modo de exploración
credito.tail()

,Trajeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
1314,yes,0,33.58333,4.566,0.002146,7.333333,yes,no,0,94,1,19.0
1315,no,5,23.91667,3.192,0.000376,0.000000,no,no,3,12,1,5.0
1316,yes,0,40.58333,4.600,0.026513,101.298300,yes,no,2,1,1,2.0
1317,yes,0,32.83333,3.700,0.008999,26.996670,no,yes,0,60,1,7.0
1318,yes,0,48.25,3.700,0.111619,344.157500,yes,no,2,2,1,0.0


In [5]:
# tamaño de los datos
credito.shape

(1319, 12)

In [6]:
# nombres de las columnas
credito.columns

Index(['Trajeta', 'Reportes Negativos', 'Edad', 'Ingreso',
       'Proporcion de Gasto', 'Gasto', 'Propietario de Vivienda',
       'Trabajor Independiente', 'Dependientes', 'Meses de Historial',
       'Tarjetas Principales', 'Cuentas Activas'],
      dtype='object')

Aca identificamos una primera cosa que podemos cambiar:

- Nombre de la primera variable

In [7]:
# cambiar el nombre de la primera variable
credito.rename(columns={"Trajeta": "Tarjeta"}, inplace=True)
credito.head()

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
0,yes,0,37.66667,4.5200,0.033270,124.983300,yes,no,3,54,1,12.0
1,yes,0,33.25,2.4200,0.005217,9.854167,no,no,3,34,1,13.0
2,yes,0,33.66667,4.5000,0.004156,15.000000,yes,no,4,5o,1,5.0
3,yes,0,30.5,2.5400,0.065214,137.869200,no,no,0,25,1,7.0
4,yes,0,32.16667,9.7867,0.067051,546.503300,yes,no,2,64,1,5.0


Ahora, veamos la información de los datos. Resumen

In [8]:
# información de los datos
credito.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Tarjeta                  1319 non-null   object 
 1   Reportes Negativos       1319 non-null   int64  
 2   Edad                     1317 non-null   object 
 3   Ingreso                  1318 non-null   float64
 4   Proporcion de Gasto      1318 non-null   float64
 5   Gasto                    1319 non-null   float64
 6   Propietario de Vivienda  1318 non-null   object 
 7   Trabajor Independiente   1317 non-null   object 
 8   Dependientes             1319 non-null   int64  
 9   Meses de Historial       1314 non-null   object 
 10  Tarjetas Principales     1319 non-null   int64  
 11  Cuentas Activas          1318 non-null   float64
dtypes: float64(4), int64(3), object(5)
memory usage: 123.8+ KB


En esta parte identificamos algunos elementos

- Algunas variables tienen datos faltantes

$$datos.isna().sum()$$

- La variable edad esta tipo object. Debe ser numérica

- Los Meses de Historial también debería ser un número, está como tipo object.

In [20]:
# para ver la cantidad de datos por columna
credito.isna().sum()

,0
Tarjeta,0
Reportes Negativos,0
Edad,2
Ingreso,1
Proporcion de Gasto,1
Gasto,0
Propietario de Vivienda,1
Trabajor Independiente,2
Dependientes,0
Meses de Historial,5


Antes de continuar con las estadísticas descriptivas vamos a solucionar los problemas anteriores.

1. Para los valores con datos faltantes:

Como son pocos datos, podríamos eliminarlos todos del dataframe con la instrucción

$$credito.dropna()$$

Sin embargo haremos dos procesos a modo de practica.

1. Cambiaremos los datos faltantes de la variable edad por el promedio.

$$datos.variable.fillna()$$

2. El restante de valores los eliminaremos

Sin embargo, hay un problema: la variable Edad es tipo *object* y así no se puede reemplazar los NA y tampoco calcular el promedio.
Debemos revisarla primero.

In [ ]:
# cambiar los datos faltantes de la variable edad de credito por el promedio de edad
credito.Edad.fillna(credito.Edad.mean(), inplace=True)
credito.info()

**Explorar la variable Edad**

In [19]:
# veamos valores únicos 0:10 revisar los 10 primeros
credito.Edad.unique()[0:10]

array(['37.66667', '33.25', '33.66667', '30.5', '32.16667', '23.25',
       '27.91667', '29.16667', '37.0', '28.41667'], dtype=object)

Como están entre comillas, es coherente con el tipo object, es decir, esta siendo tomada como cadena de texto.

Veamos donde está el problema con estos datos.

**Coerción**

La coerción es el proceso mediante el cual un lenguaje o una función convierte automáticamente un dato de un tipo a otro para poder realizar una operación.

In [23]:
# ejemplo de coerción
pd.to_numeric(credito.Edad, errors='coerce')

,Edad
0,37.66667
1,33.25000
2,33.66667
3,30.50000
4,32.16667
...,...
1314,33.58333
1315,23.91667
1316,40.58333
1317,32.83333


In [25]:
# identificamos donde hay errores
pd.to_numeric(credito.Edad, errors='coerce').isna()

,Edad
0,False
1,False
2,False
3,False
4,False
...,...
1314,False
1315,False
1316,False
1317,False


In [26]:
# identificamos donde hay errores
pd.to_numeric(credito.Edad, errors='coerce').isna().sum()

np.int64(4)

Hay 4 valores con problemas

Podemos usar la isntrucción

$$np.where()$$

para revisar en cuáles posiciones están estos problemas

In [27]:
# identificamos las posiciones
np.where(pd.to_numeric(credito.Edad, errors='coerce').isna())

(array([ 11,  15, 116, 569]),)

In [28]:
# utilizamos iloc
credito.iloc[[ 11,  15, 116, 569]]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
11,no,0,,1.98,0.000606,0.0000,yes,no,2,36,1,0.0
15,yes,0,,3.20,0.125819,335.4350,yes,no,1,168,1,22.0
116,yes,1,NaN,2.00,0.060808,101.2625,yes,no,3,42,1,3.0
569,yes,0,NaN,2.10,0.321348,562.3592,no,no,0,28,1,5.0


Observamos que hay dos valores vacíos y dos valores faltantes.

In [29]:
# ahora sí reescribo la coerción en la variable
credito["Edad"] = pd.to_numeric(credito["Edad"], errors="coerce")

In [31]:
credito.Edad.unique()[0:10]

array([37.66667, 33.25   , 33.66667, 30.5    , 32.16667, 23.25   ,
       27.91667, 29.16667, 37.     , 28.41667])

In [32]:
np.where(pd.to_numeric(credito.Edad, errors='coerce').isna())

(array([ 11,  15, 116, 569]),)

La diferencia es que la variable es tipo numérica

In [ ]:
credito.Edad[[11,  15, 116, 569]] = credito.Edad.mean()

In [35]:
credito.Edad.isna().sum()

np.int64(0)

Ya no hay datos faltantes y además

In [36]:
credito.iloc[[11,  15, 116, 569]]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
11,no,0,33.615526,1.98,0.000606,0.0000,yes,no,2,36,1,0.0
15,yes,0,33.615526,3.20,0.125819,335.4350,yes,no,1,168,1,22.0
116,yes,1,33.615526,2.00,0.060808,101.2625,yes,no,3,42,1,3.0
569,yes,0,33.615526,2.10,0.321348,562.3592,no,no,0,28,1,5.0


**veamos la otra variable**

In [38]:
credito.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Tarjeta                  1319 non-null   object 
 1   Reportes Negativos       1319 non-null   int64  
 2   Edad                     1319 non-null   float64
 3   Ingreso                  1318 non-null   float64
 4   Proporcion de Gasto      1318 non-null   float64
 5   Gasto                    1319 non-null   float64
 6   Propietario de Vivienda  1318 non-null   object 
 7   Trabajor Independiente   1317 non-null   object 
 8   Dependientes             1319 non-null   int64  
 9   Meses de Historial       1314 non-null   object 
 10  Tarjetas Principales     1319 non-null   int64  
 11  Cuentas Activas          1318 non-null   float64
dtypes: float64(5), int64(3), object(4)
memory usage: 123.8+ KB


In [40]:
# revisar cadenas de texto en months
credito["Meses de Historial"].unique()[0:10]

array(['54', '34', '5o', '25', '64', '7', '77', '97', '65', '24'],
      dtype=object)

De igual forma es cadena de texto

In [42]:
# veamos cuantos problemas hay.
pd.to_numeric(credito["Meses de Historial"], errors='coerce').isna().sum()

np.int64(8)

In [44]:
# veamos donde están los problemas
np.where(pd.to_numeric(credito["Meses de Historial"], errors='coerce').isna())[0]

array([  2,  14,  32, 252, 320, 399, 465, 496])

In [45]:
credito.iloc[[2,  14,  32, 252, 320, 399, 465, 496]]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
2,yes,0,33.66667,4.500,0.004156,15.00000,yes,no,4,5o,1,5.0
14,yes,0,35.33333,1.908,0.026907,42.61500,yes,no,2,12o,0,5.0
32,no,1,24.91667,3.150,0.000381,0.00000,yes,no,1,4o,1,5.0
252,yes,0,25.08333,2.650,0.009651,20.47917,no,no,0,NaN,1,9.0
320,no,2,28.16667,2.600,0.000462,0.00000,yes,no,2,NaN,1,3.0
399,yes,0,22.58333,2.250,0.011111,20.08333,NaN,no,0,NaN,1,4.0
465,no,0,23.83333,2.200,0.000545,0.00000,no,no,0,NaN,0,0.0
496,yes,0,42.25000,3.235,0.024766,66.26417,yes,no,2,NaN,1,3.0


In [ ]:
#credito.iloc[list(np.where(pd.to_numeric(credito.months, errors='coerce').isna())[0])]

Haremos dos cosas

- Eliminados los NA

- Cambiamos la "o" por "0".


In [46]:
# para eliminar NA's
credito.dropna(inplace=True)

In [47]:
credito.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1309 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Tarjeta                  1309 non-null   object 
 1   Reportes Negativos       1309 non-null   int64  
 2   Edad                     1309 non-null   float64
 3   Ingreso                  1309 non-null   float64
 4   Proporcion de Gasto      1309 non-null   float64
 5   Gasto                    1309 non-null   float64
 6   Propietario de Vivienda  1309 non-null   object 
 7   Trabajor Independiente   1309 non-null   object 
 8   Dependientes             1309 non-null   int64  
 9   Meses de Historial       1309 non-null   object 
 10  Tarjetas Principales     1309 non-null   int64  
 11  Cuentas Activas          1309 non-null   float64
dtypes: float64(5), int64(3), object(4)
memory usage: 132.9+ KB


In [48]:
# volvemos a revisar
np.where(pd.to_numeric(credito["Meses de Historial"], errors='coerce').isna())[0]

array([ 2, 13, 29])

In [49]:
credito.iloc[[2, 13, 29]]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
2,yes,0,33.66667,4.500,0.004156,15.000,yes,no,4,5o,1,5.0
14,yes,0,35.33333,1.908,0.026907,42.615,yes,no,2,12o,0,5.0
32,no,1,24.91667,3.150,0.000381,0.000,yes,no,1,4o,1,5.0


In [50]:
# reemplazar la o por 0 en la variable Meses de Historial
credito["Meses de Historial"] = credito["Meses de Historial"].str.replace("o", "0")

In [51]:
credito.iloc[[2, 13, 29]]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
2,yes,0,33.66667,4.500,0.004156,15.000,yes,no,4,50,1,5.0
14,yes,0,35.33333,1.908,0.026907,42.615,yes,no,2,120,0,5.0
32,no,1,24.91667,3.150,0.000381,0.000,yes,no,1,40,1,5.0


In [52]:
credito["Meses de Historial"].unique()[0:10]

array(['54', '34', '50', '25', '64', '7', '77', '97', '65', '24'],
      dtype=object)

In [53]:
# convertir a numérico
credito["Meses de Historial"] = pd.to_numeric(credito["Meses de Historial"], errors="coerce")

In [54]:
# verificamos
credito["Meses de Historial"].unique()[0:10]

array([54, 34, 50, 25, 64,  7, 77, 97, 65, 24])

**Para pensar un poco**

¿Para no estar definiendo en cada momento los mismos elemento, puedo definir algún objeto para utilizarlo en ciertos procesos?

```python
# función para detectar
def problema_texto(datos, x):
  u = pd.to_numeric(datos[x], errors='coerce').isna()
  posiciones = np.where(u)[0]
  print(posiciones)
  return datos.iloc[posiciones]
```





Ahora veamos las categorias de las variables tipo object

In [75]:
credito["Tarjeta"].value_counts()

,count
Tarjeta,
yes,1012
no,289
Yes,5
No,3


observamos que hay cuatro categorias y deben ser dos

In [77]:
# cambiar Yes y No por yes o no
#credito["Tarjeta"] = credito["Tarjeta"].str.lower()
credito["Tarjeta"] = credito["Tarjeta"].replace({"Yes": "yes", "No": "no"})

In [78]:
credito["Tarjeta"].value_counts()

,count
Tarjeta,
yes,1017
no,292


  **Para pensar un poco**

¿Cómo podemos hacer el proceso para todas las columnas?

```python
for name in credito.columns:
  if credito[name].dtype == "object":
    u=credito[name].unique()
    print(name , u)
```



Más adelante responderemos esas preguntas, por ahora, sigamos con la exploración.

In [80]:
#estadísticas descriptivas: cuantitativas
credito.describe()

,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
count,1309.000000,1309.000000,1309.000000,1309.000000,1309.000000,1309.000000,1309.000000,1309.000000,1309.000000
mean,0.453018,33.636016,3.365272,0.076818,185.843156,0.992361,55.470588,0.816654,7.029030
std,1.337063,18.818866,1.689178,0.186818,272.755147,1.245151,66.463184,0.387098,6.312994
min,0.000000,-120.000000,0.210000,0.000109,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,25.416670,2.236000,0.002424,5.520833,0.000000,12.000000,1.000000,2.000000
50%,0.000000,31.250000,2.900000,0.039573,101.854200,1.000000,30.000000,1.000000,6.000000
75%,0.000000,39.416670,4.000000,0.094789,250.520800,2.000000,72.000000,1.000000,11.000000
max,14.000000,560.000000,13.500000,3.890000,3099.505000,6.000000,540.000000,1.000000,46.000000


In [82]:
#estadísticas descriptivas: cualitativas
credito.describe(include = "object")

,Tarjeta,Propietario de Vivienda,Trabajor Independiente
count,1309,1309,1309
unique,2,4,2
top,yes,no,no
freq,1017,734,1219


Lo que sigue es revisar las estadísticas descriptivas variable por variable y revisar si los valores son coherentes.

Por ejemplo, la edad parece estar bien

In [83]:
credito.Edad.describe()

,Edad
count,1309.000000
mean,33.636016
std,18.818866
min,-120.000000
25%,25.416670
50%,31.250000
75%,39.416670
max,560.000000


Aparece una edad de 560 años, lo cual no tiene sentido.
Podríamos hacer un filtro y considerar edades mayores a 18 años y menores a 75 años de acuerdo con la naturaleza del problema

In [84]:
credito[~credito.Edad.between(18,75)]

,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
23,yes,1,200.000000,4.0000,0.051192,170.64080,no,no,2,18,1,22.0
78,yes,0,0.500000,3.0500,0.101724,258.54920,no,no,1,94,1,5.0
249,yes,0,83.500000,2.8725,0.005849,13.33333,yes,yes,0,38,0,11.0
323,yes,0,0.166667,3.2400,0.184366,497.70580,yes,no,3,25,1,16.0
358,yes,0,560.000000,10.0000,0.099808,831.73170,yes,no,3,7,1,28.0
434,yes,0,0.583333,2.5000,0.083171,173.02330,no,no,0,150,1,5.0
459,yes,1,-120.000000,4.0000,0.051054,170.18080,yes,no,5,28,1,7.0
461,no,0,0.750000,3.0000,0.000400,0.00000,no,no,0,18,0,2.0
655,yes,0,0.583333,4.0000,0.072664,242.12830,yes,no,3,24,1,4.0
658,yes,1,0.500000,3.7000,0.010637,32.46416,no,no,0,186,0,5.0


aca tomamos la decisión de solo considerar los de la mitad

In [85]:
# edades entre 18 y 75 años
credito = credito[credito.Edad.between(18,75)]

In [86]:
credito.describe()

,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
count,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000
mean,0.454896,33.324630,3.362898,0.076900,185.561819,0.988435,55.511180,0.817271,7.006168
std,1.342449,9.720094,1.684846,0.187596,273.103419,1.240030,66.541237,0.386594,6.289563
min,0.000000,18.166670,0.210000,0.000109,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,25.416670,2.234800,0.002404,4.583333,0.000000,12.000000,1.000000,2.000000
50%,0.000000,31.333330,2.900000,0.039305,101.683300,1.000000,30.000000,1.000000,6.000000
75%,0.000000,39.333330,4.000000,0.094348,250.520800,2.000000,72.000000,1.000000,11.000000
max,14.000000,74.166660,13.500000,3.890000,3099.505000,6.000000,540.000000,1.000000,46.000000


In [87]:
credito.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1297 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Tarjeta                  1297 non-null   object 
 1   Reportes Negativos       1297 non-null   int64  
 2   Edad                     1297 non-null   float64
 3   Ingreso                  1297 non-null   float64
 4   Proporcion de Gasto      1297 non-null   float64
 5   Gasto                    1297 non-null   float64
 6   Propietario de Vivienda  1297 non-null   object 
 7   Trabajor Independiente   1297 non-null   object 
 8   Dependientes             1297 non-null   int64  
 9   Meses de Historial       1297 non-null   int64  
 10  Tarjetas Principales     1297 non-null   int64  
 11  Cuentas Activas          1297 non-null   float64
dtypes: float64(5), int64(4), object(3)
memory usage: 131.7+ KB


Ya con el conjunto de datos listos podemos hacer tablas de frecuencias, y tablas cruzadas (de contingencia), etc.

# <FONT SIZE=5 COLOR="purple"> 2. Funciones en Python </FONT>

- Las funciones son bloques de código reutilizables que permiten organizar y estructurar programas de manera eficiente.

- Su principal objetivo es encapsular tareas específicas, facilitando la lectura, el mantenimiento y la reutilización del código.

- En Python, las funciones se definen mediante la palabra clave *def*
y pueden recibir parámetros de entrada, realizar operaciones y devolver resultados.

```python
def function(x,y,x..):
  .....
  return ...  
```
- El uso de funciones contribuye a reducir la duplicación de código, mejorar la modularidad de los programas y simplificar la resolución de problemas complejos mediante su división en tareas más pequeñas.-

- Favorecen la claridad del código y permiten desarrollar aplicaciones más robustas y fáciles de depurar.

En esta sección se presentan los conceptos básicos relacionados con la definición, uso y aplicación de funciones en Python, así como algunos ejemplos prácticos.

In [95]:
# ejemplo básico de función sin parámetros
def saludo1():
  print("Hola cómo están. Esto es un saludo")

cada vez que escribamos *saludo()*

In [96]:
saludo1()

Hola cómo están. Esto es un saludo


In [97]:
# ejemplo básico de función con parámetros
def saludo2(nombre):
  print("Hola" , nombre, "Cómo estás?")

In [98]:
saludo2("Fabian")

Hola Fabian Cómo estás?


**Formas diferentes del print()**

Para imprimir variables dentro del *print* hay varias alternativas

- **f-strings**

- **.format()**

**método f"{variable}"**

In [102]:
# método f"{variable}"
def saludo3(nombre):
  print(f"Hola, {nombre} espero que estes bien")

In [103]:
saludo3("Carolina")

Hola, Carolina espero que estes bien


**método "{}".format(variable)**

In [104]:
def saludo4(nombre):
    print("Hola {}, me alegra mucho verte".format(nombre))

In [105]:
saludo4("Carlos")

Hola Carlos, me alegra mucho verte


Las funciones pueden tener varios parámetros

In [115]:
def suma1(x,y):
  print(x+y)

suma1(10,89)

99


In [116]:
# veamos lo siguiente
x = suma1(10,89)

99


In [117]:
# veamos el valor de x
print(x)

None


No guarda la información!!

In [118]:
def suma2(x,y):
  return x+y

In [119]:
suma2(10,89)

99

In [121]:
#hacemos lo mismo
x=suma2(10,89)
print(x)

99


Podemos pensar como una calculadora.

- **print**:

  - la calculadora muestra el resultado en la pantalla.
  - cuando se quiere mostrar únicamente la información

- **return**:

  - la calculadora te entrega el resultado para que puedas usarlo en otro cálculo.
  - cuando queremos un resultados para seguir trabajando con él.

Veamos otros ejemplos de funciones


In [132]:
def funcion1(x,y):
  z = x^2+y**2+x*y**3+60
  return(z)

In [133]:
funcion1(10,-13)

-21729

**Ejercicio**

Generar funciones para:

- Calcular el área de un rectángulo
- Conversión de celsius a fahrenheit
- Generar un correo intitucional: nombre.apellido@universidad.edu.co
- Calculo el índice de masa corporal
- Calcular interés simple

In [158]:
def exploracion_rapida(datos):
    primeros = datos.head()
    ultimos = datos.tail()
    return primeros, ultimos

In [157]:
exploracion_rapida(credito)

(  Tarjeta  Reportes Negativos      Edad  Ingreso  Proporcion de Gasto  \
 0     yes                   0  37.66667   4.5200             0.033270   
 1     yes                   0  33.25000   2.4200             0.005217   
 2     yes                   0  33.66667   4.5000             0.004156   
 3     yes                   0  30.50000   2.5400             0.065214   
 4     yes                   0  32.16667   9.7867             0.067051   
 
         Gasto Propietario de Vivienda Trabajor Independiente  Dependientes  \
 0  124.983300                     yes                     no             3   
 1    9.854167                      no                     no             3   
 2   15.000000                     yes                     no             4   
 3  137.869200                      no                     no             0   
 4  546.503300                     yes                     no             2   
 
    Meses de Historial  Tarjetas Principales  Cuentas Activas  
 0            

In [187]:
def exploracion(datos):
    print("Primeros 5 registros \n")
    display(datos.head())

    print("\n Últimos 5 registros \n")
    display(datos.tail())

    print("\n Tamaño de los datos \n")
    print(datos.shape)
    print(f"Filas: {datos.shape[0]}")
    print(f"Columnas: {datos.shape[1]}")

    print("\n Nombres de las columnas \n")
    print(datos.columns)

    print("\n Información de los datos \n")
    print(datos.info())

    print("\n Estadísticas descriptivas \n")
    display(datos.describe())

    print("\n Valores faltantes \n")
    display(datos.isna().sum())

In [188]:
exploracion(credito)

Primeros 5 registros 



,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
0,yes,0,37.66667,4.5200,0.033270,124.983300,yes,no,3,54,1,12.0
1,yes,0,33.25000,2.4200,0.005217,9.854167,no,no,3,34,1,13.0
2,yes,0,33.66667,4.5000,0.004156,15.000000,yes,no,4,50,1,5.0
3,yes,0,30.50000,2.5400,0.065214,137.869200,no,no,0,25,1,7.0
4,yes,0,32.16667,9.7867,0.067051,546.503300,yes,no,2,64,1,5.0



 Últimos 5 registros 



,Tarjeta,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Propietario de Vivienda,Trabajor Independiente,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
1314,yes,0,33.58333,4.566,0.002146,7.333333,yes,no,0,94,1,19.0
1315,no,5,23.91667,3.192,0.000376,0.000000,no,no,3,12,1,5.0
1316,yes,0,40.58333,4.600,0.026513,101.298300,yes,no,2,1,1,2.0
1317,yes,0,32.83333,3.700,0.008999,26.996670,no,yes,0,60,1,7.0
1318,yes,0,48.25000,3.700,0.111619,344.157500,yes,no,2,2,1,0.0



 Tamaño de los datos 

(1297, 12)
Filas: 1297
Columnas: 12

 Nombres de las columnas 

Index(['Tarjeta', 'Reportes Negativos', 'Edad', 'Ingreso',
       'Proporcion de Gasto', 'Gasto', 'Propietario de Vivienda',
       'Trabajor Independiente', 'Dependientes', 'Meses de Historial',
       'Tarjetas Principales', 'Cuentas Activas'],
      dtype='object')

 Información de los datos 

<class 'pandas.core.frame.DataFrame'>
Index: 1297 entries, 0 to 1318
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Tarjeta                  1297 non-null   object 
 1   Reportes Negativos       1297 non-null   int64  
 2   Edad                     1297 non-null   float64
 3   Ingreso                  1297 non-null   float64
 4   Proporcion de Gasto      1297 non-null   float64
 5   Gasto                    1297 non-null   float64
 6   Propietario de Vivienda  1297 non-null   object 
 7   Trabajor Independient

,Reportes Negativos,Edad,Ingreso,Proporcion de Gasto,Gasto,Dependientes,Meses de Historial,Tarjetas Principales,Cuentas Activas
count,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000,1297.000000
mean,0.454896,33.324630,3.362898,0.076900,185.561819,0.988435,55.511180,0.817271,7.006168
std,1.342449,9.720094,1.684846,0.187596,273.103419,1.240030,66.541237,0.386594,6.289563
min,0.000000,18.166670,0.210000,0.000109,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,25.416670,2.234800,0.002404,4.583333,0.000000,12.000000,1.000000,2.000000
50%,0.000000,31.333330,2.900000,0.039305,101.683300,1.000000,30.000000,1.000000,6.000000
75%,0.000000,39.333330,4.000000,0.094348,250.520800,2.000000,72.000000,1.000000,11.000000
max,14.000000,74.166660,13.500000,3.890000,3099.505000,6.000000,540.000000,1.000000,46.000000



 Valores faltantes 



,0
Tarjeta,0
Reportes Negativos,0
Edad,0
Ingreso,0
Proporcion de Gasto,0
Gasto,0
Propietario de Vivienda,0
Trabajor Independiente,0
Dependientes,0
Meses de Historial,0


In [180]:
def dimensiones(df):
    return pd.DataFrame({
        "Filas": [df.shape[0]],
        "Columnas": [df.shape[1]]
    })

In [181]:
dimensiones(credito)

,Filas,Columnas
0,1297,12


In [182]:
def tipos(df):
    return pd.DataFrame({
        "Columna": df.columns,
        "Tipo": df.dtypes.values
    })

In [183]:
tipos(credito)

,Columna,Tipo
0,Tarjeta,object
1,Reportes Negativos,int64
2,Edad,float64
3,Ingreso,float64
4,Proporcion de Gasto,float64
5,Gasto,float64
6,Propietario de Vivienda,object
7,Trabajor Independiente,object
8,Dependientes,int64
9,Meses de Historial,int64


In [185]:
def valores_unicos(df, columna):
    return pd.DataFrame({
        "Valores únicos": df[columna].unique()
    })

In [186]:
valores_unicos(credito, "Tarjeta")

,Valores únicos
0,yes
1,no
